### PR TITLE
Update smeared cracking model

### PR DESCRIPTION
**Pull request by liuwenf**
- The smeared cracking modeling is improved in its computation of stresses 
  using othotropic material property; added computational method for full shear retention;
  added parameter to adjust fracture energy; added one test case under plane stress condition.

-ref.#7226

### CODE REVIEW

**Comment by dschwen** [comment]
Looks like you rebased on master...

**Comment by andrsd** [comment]
- `git checkout update_smeared_cracking_model_7226`
- `git rebase --onto devel HEAD~1 update_smeared_cracking_model_7226`

And force push to fix this...

**Comment by liuwenf** [comment]
I did the rebase as you suggested. Please check if it is on the devel branch. Thanks.

**Comment by dschwen** [comment]
> Please check if it is on the devel branch. Thanks.

You can check that yourself. Just look at the _Commits_ tab above. If it shows commits that are not yours you have screwed up. It looks good now.

File: modules/solid_mechanics/src/materials/SolidModel.C [hunk 1]
```diff
@@ -48,6 +48,8 @@ InputParameters validParams<SolidModel>()
   params.addParam<std::string>("cracking_release", "abrupt", "The cracking release type.  Choices are abrupt (default) and exponential.");
   params.addParam<Real>("cracking_stress", 0.0, "The stress threshold beyond which cracking occurs.  Must be positive.");
   params.addParam<Real>("cracking_residual_stress", 0.0, "The fraction of the cracking stress allowed to be maintained following a crack.");
+  params.addParam<Real>("cracking_beta", 1.0, "The coefficient used in the exponetional model.");
+  params.addParam<int>("compute_method", 1, "The method  used in the stress calculation.");
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/SolidModel.C]: I think this should be a `MooseEnum`

File: modules/solid_mechanics/src/materials/SolidModel.C [hunk 2]
```diff
@@ -107,6 +109,8 @@ SolidModel::SolidModel( const InputParameters & parameters) :
   _cracking_stress( parameters.isParamValid("cracking_stress") ?
                     (getParam<Real>("cracking_stress") > 0 ? getParam<Real>("cracking_stress") : -1) : -1 ),
   _cracking_residual_stress( getParam<Real>("cracking_residual_stress") ),
+  _cracking_beta( getParam<Real>("cracking_beta") ),
+  _compute_method( getParam<int>("compute_method") ),
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/SolidModel.C]: We do not put whitespace on the interior of parens.

File: modules/solid_mechanics/src/materials/SolidModel.C [hunk 3]
```diff
@@ -1122,6 +1130,7 @@ SolidModel::computeCrackStrainAndOrientation( ColumnMajorMatrix & principal_stra
 void
 SolidModel::crackingStressRotation()
 {
+
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/SolidModel.C]: remove those added empty lines, please they do not improve readability

File: modules/solid_mechanics/src/materials/SolidModel.C [hunk 4]
```diff
@@ -1122,6 +1130,7 @@ SolidModel::computeCrackStrainAndOrientation( ColumnMajorMatrix & principal_stra
...
   if (_cracking_stress_function != NULL)
    {
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/SolidModel.C]: could you fix the indentation here?

File: modules/solid_mechanics/src/materials/SolidModel.C [hunk 5]
```diff
@@ -1311,7 +1322,7 @@ SolidModel::computeCrackFactor( int i, Real & sigma, Real & flagVal )
     // Compute stress that follows exponental curve
     sigma = _cracking_stress*
            (_cracking_residual_stress + (1-_cracking_residual_stress)*
-            std::exp(_cracking_alpha/_cracking_stress*(crackMaxStrain-(*_crack_strain)[_qp](i))));
+            std::exp(_cracking_alpha * _cracking_beta /_cracking_stress*(crackMaxStrain-(*_crack_strain)[_qp](i))));
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/SolidModel.C]: more operator spacing could be fixed here

File: modules/solid_mechanics/src/materials/SymmIsotropicElasticityTensor.C [hunk 6]
```diff
@@ -231,10 +231,22 @@ SymmIsotropicElasticityTensor::multiply( const SymmTensor & x, SymmTensor & b )
...
+  b.zx() += _val[5]*xx +_val[10]*yy + _val[14]*zz;
+
+  b.yz() +=  2*_val[16] * xy;
+  b.zx() += 2*_val[17] * xy + 2*_val[19] * yz;
+
 }
 
 void
-SymmIsotropicElasticityTensor::adjustForCracking( const RealVectorValue & crack_flags )
+SymmIsotropicElasticityTensor::adjustForCracking( const RealVectorValue & crack_flags, bool hasFullShearRetention )
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/SymmIsotropicElasticityTensor.C]: `hasFullShearRetention` has no `const` qualifier here. We usually do _not_ specify `const` for pass by value (although it does have an application). I would suggest removing it everywhere for the new parameter.

File: modules/solid_mechanics/src/materials/SymmIsotropicElasticityTensor.C [hunk 7]
```diff
@@ -272,12 +284,15 @@ SymmIsotropicElasticityTensor::adjustForCracking( const RealVectorValue & crack_
   _val[13] *= c12;
   _val[14] *= c02;
 
-  _val[15] *= c01;
-  _val[16] *= c012;
-  _val[17] *= c012;
+  if ( ! hasFullShearRetention )
+    {
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/SymmIsotropicElasticityTensor.C]: indentation

File: modules/solid_mechanics/src/materials/SymmIsotropicElasticityTensor.C [hunk 8]
```diff
@@ -231,10 +231,22 @@ SymmIsotropicElasticityTensor::multiply( const SymmTensor & x, SymmTensor & b )
   b.xy() = 2*_val[15]*xy;
   b.yz() = 2*_val[18]*yz;
   b.zx() = 2*_val[20]*zx;
+
+  b.xx() += 2*(_val[3]*xy + _val[4]*yz +_val[5]*zx);
+  b.yy() += 2*(_val[8]*xy + _val[9]*yz + _val[10]*zx);
+  b.zz() += 2*(_val[12]*xy +_val[13]*yz + _val[14]*zx);
+
+  b.xy() += _val[3]*xx + _val[8]*yy +_val[12]*zz;
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/SymmIsotropicElasticityTensor.C]: operator spacing

File: modules/solid_mechanics/src/materials/SolidModel.C [hunk 9]
```diff
@@ -989,7 +994,10 @@ SolidModel::crackingStrainDirections()
   {
     // Adjust the elasticity matrix for cracking.  This must be used by the
     // constitutive law.
-    _local_elasticity_tensor->adjustForCracking( _crack_flags_local );
+    if (_compute_method >= 2)
+      _local_elasticity_tensor->adjustForCracking( _crack_flags_local, true);
+    else
+      _local_elasticity_tensor->adjustForCracking( _crack_flags_local, false);
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/SolidModel.C]: I do not like this design. Passing bool flags to methods is not a good pattern (their meaning is not clear when looking at the call (unlike python which has named parameters)). I would suggest adding a new method for the `true` case with an expressive name (`adjustForCrackingWithShearRetention`). That method could call `adjustForCracking` and then perform the necessary modifications on `_val`.

Comment (dschwen) [inline modules/solid_mechanics/src/materials/SolidModel.C]: Ugh, that's not quite that easy, as the multipliers would not be available in the new method. Never mind then. Alternative suggestion would be to pass in a custom `enum` (`WITH_SHEAR_RETENTION`, `WITHOUT_SHEAR_RETENTION`) to make the call more expressive.

File: modules/solid_mechanics/include/materials/SymmElasticityTensor.h [hunk 10]
```diff
@@ -151,7 +151,8 @@ class SymmElasticityTensor
   void rotateFromGlobalToLocal( const ColumnMajorMatrix & R );
   void rotateFromLocalToGlobal( const ColumnMajorMatrix & R );
 
-  virtual void adjustForCracking( const RealVectorValue & crack_flags );
+  virtual void adjustForCracking( const RealVectorValue & crack_flags);
+  virtual void adjustForCrackingWithShearRetention( const RealVectorValue & crack_flags);
```
Comment (dschwen) [inline modules/solid_mechanics/include/materials/SymmElasticityTensor.h]: More space on the interior of parens

File: modules/solid_mechanics/src/materials/SolidModel.C [hunk 11]
```diff
@@ -107,6 +110,8 @@ SolidModel::SolidModel( const InputParameters & parameters) :
   _cracking_stress( parameters.isParamValid("cracking_stress") ?
                     (getParam<Real>("cracking_stress") > 0 ? getParam<Real>("cracking_stress") : -1) : -1 ),
   _cracking_residual_stress( getParam<Real>("cracking_residual_stress") ),
+  _cracking_beta(getParam<Real>("cracking_beta")),
+  _compute_method(getParam<MooseEnum>("compute_method") == "ShearRetention" ? 2 : 1),
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/SolidModel.C]: No. MooseEnums evaluate to integers. This is an un-pattern!

File: modules/solid_mechanics/src/materials/SolidModel.C [hunk 12]
```diff
@@ -989,7 +995,10 @@ SolidModel::crackingStrainDirections()
   {
     // Adjust the elasticity matrix for cracking.  This must be used by the
     // constitutive law.
-    _local_elasticity_tensor->adjustForCracking( _crack_flags_local );
+    if (_compute_method >= 2)
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/SolidModel.C]: No. Why are you using `>=` here?

File: modules/solid_mechanics/src/materials/SymmElasticityTensor.C [hunk 13]
```diff
@@ -363,6 +363,11 @@ SymmElasticityTensor::adjustForCracking( const RealVectorValue & /*crack_flags*/
 }
 
 void
+SymmElasticityTensor::adjustForCrackingWithShearRetention( const RealVectorValue & /*crack_flags*/)
+{
+  mooseError("adjustForCrackingWithShearRetention method not defined");
+}
+void
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/SymmElasticityTensor.C]: empty line between functions please

File: modules/solid_mechanics/src/materials/SymmIsotropicElasticityTensor.C [hunk 14]
```diff
@@ -231,10 +231,22 @@ SymmIsotropicElasticityTensor::multiply( const SymmTensor & x, SymmTensor & b )
...
+  b.zx() += _val[5]*xx +_val[10]*yy + _val[14]*zz;
+
+  b.yz() +=  2*_val[16] * xy;
+  b.zx() += 2*_val[17] * xy + 2*_val[19] * yz;
+
 }
 
 void
-SymmIsotropicElasticityTensor::adjustForCracking( const RealVectorValue & crack_flags )
+SymmIsotropicElasticityTensor::adjustForCracking( const RealVectorValue & crack_flags)
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/SymmIsotropicElasticityTensor.C]: no space on the inside of parens. You have this all over the place.

File: modules/solid_mechanics/src/materials/SolidModel.C (hunk 12 above)
Comment (liuwenf) [inline modules/solid_mechanics/src/materials/SolidModel.C]: To accommodate a new method under development.

Comment (dschwen) [inline modules/solid_mechanics/src/materials/SolidModel.C]: Uhm, no. If you implement a new method you will surely have to adjust this if as well, right? Might as well make it a switch statement with a `default:` clause that throws a mooseError

File: modules/solid_mechanics/include/materials/SolidModel.h [hunk 15]
```diff
@@ -86,6 +86,8 @@ class SolidModel : public DerivativeMaterialInterface<Material>
   const CRACKING_RELEASE _cracking_release;
   Real _cracking_stress;
   const Real _cracking_residual_stress;
+  const Real _cracking_beta;
+  const std::string  _compute_method;
```
Comment (dschwen) [inline modules/solid_mechanics/include/materials/SolidModel.h]: Why did you make this a string now? Leave it an int. (and remove that extra space)

File: modules/solid_mechanics/src/materials/SymmIsotropicElasticityTensor.C [hunk 16]
```diff
@@ -225,16 +225,27 @@ SymmIsotropicElasticityTensor::multiply( const SymmTensor & x, SymmTensor & b )
...
+  b.yz() = 2 * _val[18] * yz;
+  b.zx() = 2 * _val[20] * zx;
+
+  b.xx() += 2 * (_val[3] * xy +  _val[4] * yz + _val[5] * zx);
+  b.yy() += 2 * (_val[8] * xy +  _val[9] * yz + _val[10] * zx);
+  b.zz() += 2 * (_val[12] * xy + _val[13] * yz + _val[14] * zx);
+
+  b.xy() += _val[3] * xx + _val[8] * yy + _val[12] * zz;
+  b.yz() += _val[4] * xx + _val[9] * yy + _val[13] * zz;
+  b.zx() += _val[5] * xx + _val[10] * yy +  _val[14] * zz;
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/SymmIsotropicElasticityTensor.C]: remove extra space

File: modules/solid_mechanics/src/materials/SymmIsotropicElasticityTensor.C [hunk 17]
```diff
@@ -272,12 +283,29 @@ SymmIsotropicElasticityTensor::adjustForCracking( const RealVectorValue & crack_
   _val[13] *= c12;
   _val[14] *= c02;
 
+}
+
+void
+SymmIsotropicElasticityTensor::adjustForCrackingWithShearRetention(const RealVectorValue & crack_flags)
+{
+  const RealVectorValue & c(crack_flags);
+  const Real c0(c(0));
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/SymmIsotropicElasticityTensor.C]: we initialize `Real` with `=` and _not_ the `()` syntax.

**Pull request merged**